### ISSUE TITLE
MinGW/MSYS compatible binary wheels

### ISSUE THREAD

**Issue by whitequark** [bytecodealliance/wasmtime-py#10]
Unexpectedly, the existing wasmtime binary wheels [don't work](https://twitter.com/cr1901/status/1263377491575660544) on MinGW/MSYS 64-bit Python. It seems to me that, since wasmtime-py doesn't depend on the Python ABI, they ought to work, but I have no deep understanding of the machinery involved.

**Comment by alexcrichton** [comment]
Unfortunately a `--plat-name mingw` [wasn't supported](https://github.com/bytecodealliance/wasmtime-py/runs/696929883?check_suite_focus=true#step:8:30) to get uploaded to pypi. I don't know how these wheels are supposed to be named to get installed on MinGW or how the platform filters work in pypi. This'll require some more investigation.

**Comment by cr1901** [comment]
@alexcrichton `mingw` isn't a supported tag on PyPI. In general, using `pip` with the `mingw` ABI python only succeeds if the package to be installed has `any` in the [platform tag](https://www.python.org/dev/peps/pep-0425/#id13) of the wheel filename. There's a way to print out all the supported tags for a given Python installation, but I forget.

If using the `none` ABI tag is unacceptable, I think a workaround for now is to force `mingw` ABI users to clone the repo and do the `PROD=1 python setup.py install` dance. And to add `mingw` API `wasmtime` to the `dev` releases so I don't have to modify `download_wasmtime.py` myself :P:

```
$ git diff download-wasmtime.py
diff --git a/download-wasmtime.py b/download-wasmtime.py
index da1491d..9af323f 100644
--- a/download-wasmtime.py
+++ b/download-wasmtime.py
@@ -13,14 +13,14 @@ def main():
     if sys.platform == 'linux':
         filename = 'wasmtime-dev-x86_64-linux-c-api.tar.xz'
     elif sys.platform == 'win32':
-        filename = 'wasmtime-dev-x86_64-windows-c-api.zip'
+        filename = 'wasmtime-v0.16.0-x86_64-mingw-c-api.zip'
         is_zip = True
     elif sys.platform == 'darwin':
         filename = 'wasmtime-dev-x86_64-macos-c-api.tar.xz'
     else:
         raise RuntimeError("unknown platform: " + sys.platform)

-    url = 'https://github.com/bytecodealliance/wasmtime/releases/download/dev/'
+    url = 'https://github.com/bytecodealliance/wasmtime/releases/download/v0.16.0/'
     url += filename
     print('Download', url)
```

**Comment by alexcrichton** [comment]
Ah thanks for the info!

One thing I was thinking of doing was to publish an `any` wheel on pypi which actually contains all of the dynamic libraries, not just the dynamic library for one platform. The extension would then select the appropriate one at load time or return an error.

@cr1901 would something like this work? Is it possible to publish *both* platform-specific and generic extensions? In that case if you install from a matching platform do you get the platform-specific version or the generic one? 

(ideally existing users would continue to only download one dynamic library, and only platforms without filters on pypi would download more-than-one binary)

**Comment by cr1901** [comment]
>would something like this work?

This would be personally fine for me.

>Is it possible to publish both platform-specific and generic extensions? In that case if you install from a matching platform do you get the platform-specific version or the generic one?

I would have to look at the pip source code or ask for clarification on the relevant mailing list (... _which mailing list would that be?_ ._.) to get an answer to these q's. I'm not sure if there's "prior art" for this scenario.

Some extra info:

Most software _I_ personally use on Windows explicitly supports the GNU ABI (often provided via a `mingw` distribution, like I assume `wasmtime-*-mingw-c-api` does<sup>1</sup>); Python is the only big piece of software I use (and that I can think of) that officially considers GNU ABI on Windows out of scope.

Often times, for packages that require special casing for `mingw`, the solution is to patch and submit to the [`mingw-packages`](https://github.com/msys2/MINGW-packages) repo upstream. However, it's a dice roll whether your package gets accepted (mine haven't ;)...). And package maintainers will [sometimes accept](https://github.com/mhammond/pywin32/pull/1371) patches for `mingw` support in their repos anyway. All this is to say: since `wasmtime` is written in Rust, and the GNU ABI is explicitly supported in Rust, it seems reasonable to support `mingw` as first-class in `wasmtime-py` even if upstream Python doesn't?

1. Fun fact: At the very least, `wasmtime-*-windows-c-api` does _not_ immediately crash when I use it with `mingw` ABI python. At least in the C ABI there's some level of compatibility between GNU and MS's ABI, but Idk where it begins and ends.

**Comment by alexcrichton** [comment]
Ah yeah so to clarify the MinGW vs MSVC ABI isn't relevant for this implementation. We just use `ctypes` and load the DLL compiled for x86_64-pc-windows-msvc in Rust. That same strategy works for both the native python and the MinGW python. I'm just trying to figure out how we ship this to users!

I think I'll try to do some testing publishing architecture-specific and architecture-generic packages and see how far I get. That'll take a bit more time, though.

**Comment by alexcrichton** [comment]
Ok I've done some testing. I'm going to tweak how wheels are published here and I think this can be fixed.

It looks like you can publish specific wheels (`--plat-name manylinux1-x86_64`) as well as a non-specific wheel (no `--plat-name` argument). Locally `pip3` seems to prefer the specific wheel over the `any` wheel if the platform matches.

With this I think I'll publish a specific wheel for each platform supported (Windows/Mac/Linux x86_64) as well as a catch-all wheel with other architectures that don't have a `--plat-name` argument. The catch-all wheel would have a precompiled copy of libwasmtime.so for all platforms other than the 3 major ones. It'd then also provide a runtime error for "your platform isn't supported".

**Comment by alexcrichton** [comment]
Ok this should be implemented in https://github.com/bytecodealliance/wasmtime-py/pull/23

**Comment by cr1901** [comment]
@alexcrichton Is there any possibility you could upload a `v0.16.2` wheel to PyPI so I can test the `any` wheel on my end?

(If there's a better way to test pip picking up the correct package from PyPI without doing a release, I'm not aware of it.)

**Comment by alexcrichton** [comment]
For testing this is continuously published to https://test.pypi.org/project/wasmtime/. For a point release there's been some breaking API changes so I'm currently waiting on 0.17.0 of wasmtime itself.
